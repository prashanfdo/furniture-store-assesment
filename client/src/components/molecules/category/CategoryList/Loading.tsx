import ContentLoader from "react-content-loader";

const Loading = () => {
  return (
    <ContentLoader
      speed={2}
      width={"100%"}
      height={150}
      backgroundColor="#f3f3f3"
      foregroundColor="#ecebeb"
      data-testid="loading"
    >
      <rect x="0" y="0" rx="5" ry="5" width="180" height="20" />
      <rect x="0" y="30" rx="5" ry="5" width="180" height="20" />
      <rect x="0" y="60" rx="5" ry="5" width="180" height="20" />
      <rect x="0" y="90" rx="5" ry="5" width="180" height="20" />
    </ContentLoader>
  );
};

export default Loading;
