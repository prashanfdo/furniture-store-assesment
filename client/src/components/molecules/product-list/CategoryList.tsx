import { css } from "@emotion/react";
import styled from "@emotion/styled";
import { ChildCategory } from "types";

type Category = ChildCategory["list"][0];

const DivContainer = styled.div`
  background-color: #fff;
`;
const H2Title = styled.h2`
  font-size: 1rem;
  padding: 0 1.5em;
`;
const UlContainer = styled.ul`
  list-style-type: none;
  margin-block-start: 0;
  margin-block-end: 0;
  margin-inline-start: 0;
  margin-inline-end: 0;
  padding-inline-start: 0;
`;
type LinkCategoryProps = {
  isActive?: boolean;
};
const LinkCategory = styled.a<LinkCategoryProps>`
  text-decoration: none;
  padding: 0.5em 1.5em;
  display: block;
  color: #000;
  transition: background-color 0.05s ease-in-out;
  ${({ isActive }) => !!isActive && css({ color: "#7047eb" })}
  font-weight: 500;
  font-size: 0.9rem;
  &:hover {
    background-color: #e6e6ea;
  }
`;

type CategoryListProps = {
  selectedCategoryId?: string;
  categories?: Category[];
};
export default function CategoryList({ categories, selectedCategoryId }: CategoryListProps) {
  return (
    <DivContainer>
      <H2Title>Categories</H2Title>
      <UlContainer>
        <li>
          <LinkCategory isActive={!selectedCategoryId?.trim()} href={"/"}>
            All
          </LinkCategory>
        </li>
        {categories?.map((category) => (
          <li key={category.urlPath}>
            <LinkCategory
              isActive={category.id === selectedCategoryId?.trim()}
              href={`/${category.id}`}>
              {category.name}
            </LinkCategory>
          </li>
        ))}
      </UlContainer>
    </DivContainer>
  );
}
