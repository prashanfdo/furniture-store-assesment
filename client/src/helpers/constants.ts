export const ROOT_CATEGORY_ID = "156126";
export const DEFAULT_LOCALE = "de_DE";
export const DEFAULT_LOCALE_NUMBER_FORMAT = DEFAULT_LOCALE.replace("_", "-");
