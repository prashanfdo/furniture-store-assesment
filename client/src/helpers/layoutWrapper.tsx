import React from "react";

type LayoutProps = {
  children: React.ReactNode;
};
type ComponentProps = {
  children?: React.ReactNode;
};
const layoutWrapper = (LayoutComponent: React.FC<LayoutProps>) => {
  return (WrappedComponent: React.FC<ComponentProps>) => {
    return class extends React.Component {
      render() {
        return (
          <LayoutComponent>
            <WrappedComponent {...this.props} />
          </LayoutComponent>
        );
      }
    };
  };
};

export default layoutWrapper;
