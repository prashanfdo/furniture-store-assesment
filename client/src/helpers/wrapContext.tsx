import React, { ReactNode } from "react";

const wrapContext = <
  TProviderProps extends { children?: ReactNode },
  TComponentProps extends { children?: ReactNode }
>(
  ContextProvider: React.FC<TProviderProps>,
  WrappedComponent: React.FC<TComponentProps>
) => {
  return (props: TProviderProps & TComponentProps) => {
    return (
      <ContextProvider {...(props as TProviderProps)}>
        <WrappedComponent {...(props as TComponentProps)} />
      </ContextProvider>
    );
  };
};

export default wrapContext;
