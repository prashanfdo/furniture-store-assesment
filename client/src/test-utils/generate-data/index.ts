export * from "./article";
export * from "./category";
export * from "./search";
