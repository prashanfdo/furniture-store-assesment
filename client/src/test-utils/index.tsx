import { render, RenderOptions } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import { BrowserRouter } from "react-router-dom";
import user from "@testing-library/user-event";

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) => {
  const view = render(ui, { wrapper, ...options });

  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};
export * from "@testing-library/react";

export { faker } from "@faker-js/faker";
export * from "./generate-data";

export { customRender as render };
export { user };
