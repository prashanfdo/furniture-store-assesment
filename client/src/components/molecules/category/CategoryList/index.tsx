import { Bars3Icon } from "@heroicons/react/20/solid";
import clsx from "clsx";
import { CategorySidebarLink } from "components/atoms";
import { ROOT_CATEGORY_ID } from "helpers/constants";
import { useState } from "react";
import { ChildCategory } from "types";

type Category = ChildCategory["list"][0];

type CategoryListProps = {
  selectedCategoryId?: string;
  categories?: Category[];
  loading: boolean;
  error: boolean;
};
const CategoryList: React.FC<CategoryListProps> = ({
  categories = [],
  selectedCategoryId,
  loading,
  error,
}: CategoryListProps) => {
  const [showCategorySlideMenu, setShowCategorySlideMenu] = useState(false);

  const handleMenuButton = () => {
    setShowCategorySlideMenu(!showCategorySlideMenu);
  };

  const isSelected = (id: string) => {
    return id === selectedCategoryId;
  };

  return (
    <>
      <aside
        className={clsx(
          "thin-scroll fixed bottom-0 left-0 lg:!left-0 w-full md:w-[220px] top-16 overflow-auto z-10 bg-white",
          {
            "!-left-full": !showCategorySlideMenu,
          },
        )}
      >
        <button
          onClick={handleMenuButton}
          className="fixed opacity-50 cursor-pointer lg:hidden top-4 left-4 hover:opacity-100"
        >
          <Bars3Icon className="w-6 h-6" />
        </button>
        <div>
          <ul>
            <li data-testid={`cat-list-item-${ROOT_CATEGORY_ID}`}>
              <CategorySidebarLink isActive={isSelected(ROOT_CATEGORY_ID)} to={"/"}>
                All
              </CategorySidebarLink>
            </li>
            {loading && <li>loading...</li>}
            {error && <li>Error Occurred</li>}
            {!loading &&
              categories?.map((category) => (
                <li key={category.id} data-testid={`cat-list-item-${category.id}`}>
                  <CategorySidebarLink
                    isActive={!!category.id ? selectedCategoryId === category.id : false}
                    to={`/${category.id}`}
                  >
                    {category.name}
                  </CategorySidebarLink>
                </li>
              ))}
          </ul>
        </div>
      </aside>
    </>
  );
};

export default CategoryList;
