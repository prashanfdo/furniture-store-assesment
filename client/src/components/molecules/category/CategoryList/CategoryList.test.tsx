import user from "@testing-library/user-event";
import { render, screen, within } from "test-utils";
import { createCategoryList } from "test-utils/generate-data";
import CategoryList from "./index";
import { ROOT_CATEGORY_ID } from "helpers/constants";

const dataCategories: React.ComponentProps<typeof CategoryList>["categories"] = createCategoryList(10);

describe("CategoryList", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const { container } = render(<CategoryList categories={dataCategories} selectedCategoryId={ROOT_CATEGORY_ID} />);
    expect(container).toMatchSnapshot();
  });

  it("should render all links", () => {
    render(<CategoryList categories={dataCategories} selectedCategoryId={ROOT_CATEGORY_ID} />);
    expect(screen.queryAllByRole("link")).toHaveLength(dataCategories.length + 1); // +1 for "All" link
  });

  it("should render home link", () => {
    render(<CategoryList categories={dataCategories} selectedCategoryId={ROOT_CATEGORY_ID} />);
    const homeLink = within(screen.getByTestId(`cat-list-item-${ROOT_CATEGORY_ID}`)).getByRole("link");
    expect(homeLink).toBeInTheDocument();
    expect(homeLink).toHaveAttribute("href", "/");
  });
});
