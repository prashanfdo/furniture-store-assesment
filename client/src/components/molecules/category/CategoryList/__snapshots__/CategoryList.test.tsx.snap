// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryList should render 1`] = `
<div>
  <aside
    class="thin-scroll fixed bottom-0 left-0 lg:!left-0 w-full md:w-[220px] top-16 overflow-auto z-10 bg-white !-left-full"
  >
    <button
      class="fixed opacity-50 cursor-pointer lg:hidden top-4 left-4 hover:opacity-100"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M2 4.75A.75.75 0 012.75 4h14.5a.75.75 0 010 1.5H2.75A.75.75 0 012 4.75zM2 10a.75.75 0 01.75-.75h14.5a.75.75 0 010 1.5H2.75A.75.75 0 012 10zm0 5.25a.75.75 0 01.75-.75h14.5a.75.75 0 010 1.5H2.75a.75.75 0 01-.75-.75z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div>
      <ul>
        <li
          data-testid="cat-list-item-156126"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all !text-[#7047eb]"
            href="/"
          >
            All
          </a>
        </li>
        <li
          data-testid="cat-list-item-41703"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/41703"
          >
            Industrial
          </a>
        </li>
        <li
          data-testid="cat-list-item-93256"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/93256"
          >
            Books
          </a>
        </li>
        <li
          data-testid="cat-list-item-30234"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/30234"
          >
            Industrial
          </a>
        </li>
        <li
          data-testid="cat-list-item-23609"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/23609"
          >
            Music
          </a>
        </li>
        <li
          data-testid="cat-list-item-18627"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/18627"
          >
            Tools
          </a>
        </li>
        <li
          data-testid="cat-list-item-66975"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/66975"
          >
            Tools
          </a>
        </li>
        <li
          data-testid="cat-list-item-53882"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/53882"
          >
            Sports
          </a>
        </li>
        <li
          data-testid="cat-list-item-31328"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/31328"
          >
            Clothing
          </a>
        </li>
        <li
          data-testid="cat-list-item-20446"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/20446"
          >
            Grocery
          </a>
        </li>
        <li
          data-testid="cat-list-item-22958"
        >
          <a
            class="decoration-0 px-6 py-2 text-sm font-semibold text-gray-800 hover:bg-slate-200 block duration-75 transition-all"
            href="/22958"
          >
            Books
          </a>
        </li>
      </ul>
    </div>
  </aside>
</div>
`;
